pkg/miner/utxpool/transaction_bulk_validator.go

```diff
@@ -11,66 +12,86 @@ import (
-		tm:    tm,
+		state:  state,
+		utx:    utx,
+		tm:     tm,
+		scheme: scheme,
 	}
 }
 
-func (a bulkValidator) Validate() {
-	transactions, err := a.validate()
-	if err != nil {
-		slog.Debug("Validation failure", logging.Error(err))
-		return
-	}
+func (a bulkValidator) Validate(ctx context.Context) {
+	transactions := a.validate(ctx) // Pop transactions from UTX, clean UTX
 	for _, t := range transactions {
-		_ = a.utx.AddWithBytes(t.T, t.B)
+		errAdd := a.utx.AddWithBytesRaw(t.T, t.B)
+		if errAdd != nil {
+			// can happen because other nodes can send copies of txs while they are being validated
+			slog.Warn("Failed to add a validated transaction to UTX",
+				logging.Error(errAdd), logging.TxID(t.T, a.scheme),
+			)
+		}
 	}
 }
 
-func (a bulkValidator) validate() ([]*types.TransactionWithBytes, error) {
-	if a.utx.Count() == 0 {
-		return nil, nil
+func (a bulkValidator) validate(ctx context.Context) []*types.TransactionWithBytes {
+	utxLen := a.utx.Count()
+	if utxLen == 0 {
+		slog.Debug("UTX pool is empty, nothing to validate")
+		return nil
 	}
 	var transactions []*types.TransactionWithBytes
 	currentTimestamp := proto.NewTimestampFromTime(a.tm.Now())
 	lastKnownBlock := a.state.TopBlock()
-
-	_ = a.state.Map(func(s state.NonThreadSafeState) error {
-		defer s.ResetValidationList()
-
-		for {
-			t := a.utx.Pop()
-			if t == nil {
-				break
-			}
-			_, err := s.ValidateNextTx(t.T, currentTimestamp, lastKnownBlock.Timestamp, lastKnownBlock.Version, false)
-			if stateerr.IsTxCommitmentError(err) {
-				// This should not happen in practice.
-				// Reset state, return applied transactions to UTX.
-				s.ResetValidationList()
-				for _, tx := range transactions {
-					_ = a.utx.AddWithBytes(tx.T, tx.B)
+	for i := range utxLen {
+		if ctx.Err() != nil {
+			slog.Debug("Bulk validation interrupted:", logging.Error(context.Cause(ctx)))
+			return transactions
+		}
+		t := a.utx.Pop()
+		if t == nil {
+			slog.Debug("UTX pool is empty, finished validating",
+				slog.Int("validated", len(transactions)),
+				slog.Int("totalTxValidated", i),
+			)
+			break
+		}
+		err := a.state.TxValidation(func(validation state.TxValidation) error {
+			_, err := validation.ValidateNextTx(t.T, currentTimestamp, lastKnownBlock.Timestamp, lastKnownBlock.Version, false)
+			return err
+		})
+		if stateerr.IsTxCommitmentError(err) {
+			slog.Error("Failed to unpack a transaction from utx", logging.Error(err))
+			// This should not happen in practice.
+			// Reset state, return applied transactions to UTX.
+			for _, tx := range transactions {
+				utxErr := a.utx.AddWithBytesRaw(tx.T, tx.B)
+				if utxErr != nil {
+					slog.Error("Failed to return a transaction to UTX",
+						logging.Error(utxErr), logging.TxID(t.T, a.scheme),
+					)
 				}
-				transactions = nil
-				continue
-			} else if err == nil {
-				transactions = append(transactions, t)
 			}
+			slog.Debug("Validated transactions returned to UTX, resetting validated list, continuing",
```

Comment:
I think the message "Valid transactions returned to UTX; resetting list and continuing" will be clear. Also rename "totalTxValidated" to "validated" or "checked".